Load basic building data

esbudylin opened the pull request:
This commit implements loading basic building data from save and scenario files. It loads the list of possible buildings; however, loading data on already existing buildings is not yet implemented.

Additionally, this commit implements loading of the currently produced units and buildings from save files.

Next, I intend to work on loading existing buildings and adding logic for creating new ones.

esbudylin committed:
Load basic building data

This commit implements loading basic building data from save and
scenario files. It loads the list of possible buildings; however,
loading data on already existing buildings is not yet implemented.

Additionally, this commit implements loading of the currently produced
units and buildings from save files.

TomWerner commented on C7GameData/Building.cs:
```diff
@@ -0,0 +1,7 @@
+	public class Building : IProducible {
+		public string name { get; set; }
+		public int shieldCost { get; set; }
+		public int populationCost { get; set; }
```
Could you leave a comment explaining what population cost is for buildings?

TomWerner replied:
Ah, I see you left a comment in `ImportCiv3.cs` - could you add that here too? Is it possible to mark this as `const` with a value of 0 to make it clearer this is to just satisfy the interface?

esbudylin replied:
I don't think an interface member can be declared as a constant. In general, I think the purpose of this interface is to allow buildings to have a population cost in modded rule sets, so I don't think this field needs to be restricted. I added a comment to clarify a situation with Civ3 rules.

TomWerner commented on C7GameData/Save/SaveGame.cs:
```diff
@@ -189,6 +190,7 @@ public GameData ToGameData() {
 		public List<Resource> Resources = new List<Resource>();
 		public List<SaveUnit> Units = new List<SaveUnit>();
 		public List<UnitPrototype> UnitPrototypes = new List<UnitPrototype>();
+		public List<Building> Buildings = new();
```
If you run the tests locally then in `C7GameDataTests/data/output/gotm_save_0.json` you should be able to grab the list of buildings from the base game rules. Could you copy paste those into `c7-static-map-save.json`?

esbudylin replied:
Thanks for the pointer! I've added the building data to the save.

TomWerner reviewed (changes requested):
Really nice!

esbudylin committed:
Fix scenario loading

esbudylin committed:
Fix sample save test

Previously, the test saved the sample save in data/ directory but then
tried to look for in data/saves directory.

TomWerner commented:
I'll get #548 merged, I fixed the goofy test issue there.

esbudylin committed:
Merge branch 'Development' into yegor/load-building-data

esbudylin committed:
Record building data into a save

esbudylin committed:
Add building data to JSON save

esbudylin committed:
Comment for populationCost field

TomWerner commented on C7/Text/c7-static-map-save.json:
```diff
@@ -61515,6 +61515,423 @@
+  "buildings": [
+    {
+      "name": "Palace",
+      "shieldCost": 100,
```
Not for this PR, but we'll need to figure out how this works, because the palace cost increases over time or per city, right?

esbudylin replied:
There is a write-up on the palace cost on Civfanatics https://forums.civfanatics.com/threads/cost-of-the-palace-for-c3c.107789/ It looks like the the palace cost depends on the number of cities.

TomWerner commented on C7/Text/c7-static-map-save.json:
```diff
@@ -64533,87 +64950,5 @@
       ]
     }
-  ],
-  "citizenTypes": [
```
Looks like these got lost in a merge, could you restore them?

TomWerner commented on C7GameData/ImportCiv3.cs:
```diff
@@ -541,6 +545,17 @@ private void ImportSavCities() {
+			PRTO[] unitPrototypes = biq.Prto ?? defaultBiq.Prto;
+
+			return city.ConstructingType switch {
+				0 => ("Warrior", ProducibleType.UNIT), // TODO: Wealth production is not implemented yet
```
Maybe make this "Worker"? Warriors aren't always producible

esbudylin committed:
Lost citizenTypes data

esbudylin committed:
Produce Worker as a placeholder for wealth
